### ISSUE TITLE
[RFC] Remove capability claim signing

### ISSUE THREAD

**Issue by brooksmtownsend** [wasmCloud/wasmCloud#1219]
# RFC: Remove capability claim signing

## Introduction
wasmCloud, before supporting Wasm components, operated on an opaque protocol called wasmbus in order for Wasm guests to interact with the wasmCloud host. In that protocol we simply wrote bytes back and forth in the form of an `Invocation`, and the invocation was destined either for an actor or a capability provider. Core to this is our idea of deny-by-default security which required that an actor was properly authorized with the capability to receive/send that invocation, as well as a runtime link to connect that actor to a capability provider.

Now that we are able to support Wasm components, capability invocations are no longer operating on an opaque protocol. Rather, they are invoked via specific Wasm imports and exports, and are specified in the WIT that's embedded in a Wasm component. Take this specific component for example that is both signed and has an embedded wit:
```

                                                                            
                          http-dog-fetcher - Actor                          
  Account         ABJXAVEVSVC5GX55IXNHWCYVNR6I6TRNA3S7KMBCZIJABLSENOL2SUYD  
  Actor           MAGZGGEO6UTISILBZVP54AWHYYSMC5VNYXZC3SM4CJLNI3OH4GODTSX3  
  Expires                                                            never  
  Can Be Used                                                  immediately  
  Version                                                        0.1.0 (0)  
  Call Alias                                                     (Not set)  
                                Capabilities                                
  HTTP Server                                                               
  Logging                                                                   
  Blob Store                                                                
  Number Generation                                                         
  HTTP Client                                                               
                                    Tags                                    
  wasmcloud.com/experimental                                                
                                                                            



package root:component;

world root {
  import wasi:io/poll@0.2.0-rc-2023-11-10;
  import wasi:io/error@0.2.0-rc-2023-11-10;
  import wasi:io/streams@0.2.0-rc-2023-11-10;
  import wasi:http/types@0.2.0-rc-2023-12-05;
  import wasi:http/outgoing-handler@0.2.0-rc-2023-12-05;
  import wasi:blobstore/types;
  import wasi:blobstore/container;
  import wasi:blobstore/blobstore;
  import wasi:logging/logging;
  import wasi:cli/environment@0.2.0-rc-2023-12-05;
  import wasi:cli/exit@0.2.0-rc-2023-12-05;
  import wasi:cli/stdin@0.2.0-rc-2023-12-05;
  import wasi:cli/stdout@0.2.0-rc-2023-12-05;
  import wasi:cli/stderr@0.2.0-rc-2023-12-05;
  import wasi:clocks/wall-clock@0.2.0-rc-2023-11-10;
  import wasi:filesystem/types@0.2.0-rc-2023-11-10;
  import wasi:filesystem/preopens@0.2.0-rc-2023-11-10;

  export wasi:http/incoming-handler@0.2.0-rc-2023-12-05;
}
```

We can see this actor was required to be signed with claims, which include blobstore, HTTP server and client, and logging. If you look at the `wit` imports and exports, we also see those exact WASI capabilities specified there. Additionally, to run in the wasmCloud ecosystem, we require runtime links to capability providers, which is a platform operator allowing those two components to talk to each other.

**This RFC suggests that we should remove capability claims from the embedded JSON token, as we already have more information from a component's WIT and we are authorizing the component to use that capability at runtime**

## Motivation

Most of the motivation is captured in #1119, which is a large RFC focused on improving developer experience and removing layers of wasmCloud specific requirements when they are no longer needed thanks to the advancement of Wasm standards. The primary motivation for this RFC is a simplification of the signing process, preventing users from the frustrating scenario of building a component, specifically telling that component to `import wasi:blobstore/container;` and then have that denied at runtime after it's already linked because it was not signed with `wasmcloud:blobstore`, a totally separate capability.

## Detailed Design

The work here should begin in the host, simply removing the checks that we have on capability claims when performing an invocation. Instead, checking that an actor component is linked to the capability provider should be enough proof that it's allowed to invoke that capability, which is already a check that we have in place.

Once that's complete, we can remove the concepts of capability claim signing from our documentation and instead replace that with noting the inspectability of a WIT world and the fact that our host has to satisfy an exact set of exports, any additional will be ignored and would fail to instantiate the component entirely.

## Backwards Compatibility

As this functionality is core to the `wash claims sign` process and `wash build`, we can first start by deprecating and hiding each of the capability claims options. This will let all existing scripts work as-is, and make each of these CLI flags a no-op, but preserve the functionality of letting the actor use that capability at runtime.

## Alternatives Considered

We could allow people to specify claims if they wanted and continue to deny invalid capabilities at runtime, which would allow more layers of security. This wasn't considered more as we are already requiring people to link to valid capability providers at runtime.

We could simply change our claims signing from `wasmcloud:*` interfaces to `wasi:*` interfaces, but again this is just repeating the same information that's discoverable from a component's world.

## Conclusion

Removing the requirement to sign capabilities in a world (heh) of components just makes the signing process simpler and less bespoke to wasmCloud, without sacrificing security.

**Comment by connorsmith256** [comment]
100% agreed. Now that the built component itself declares which capabilities it imports, there's no need to require signing it with a capability. Controlling access to capabilities at runtime is handled just fine via runtime links 👍

**Comment by thomastaylor312** [comment]
Also entirely in favor of this. Thanks to @autodidaddict for all the initial work and thinking here and @brooksmtownsend for getting this together!

**Comment by vados-cosmonic** [comment]
Capability signing is removed by https://github.com/wasmCloud/wasmCloud/pull/1748, and now the docs have been updated as well. 

`wasmCloud/wasmCloud` code has also been updated, and while there is some code in `wasmCloud/examples` that probably could be updated, we're going to be moving those into `wasmCloud/wasmCloud` (and we haven't released 1.0 yet) so marking this as done.
